elastic/kibana
# [ResponseOps][Alerting] Prevent task state validation failures from failing a task claiming cycle

Resolves https://github.com/elastic/kibana/issues/170337

## Summary

Currently Task Manager can claim a task with an invalid state schema and fail all future task claiming cycles. In this PR we move the task state validation up to higher level and fail the task run instead of preventing all tasks from running.

### Checklist



### To verify

1. In your kibana.yml, set `xpack.task_manager.allow_reading_invalid_state: false`
2. Startup Elasticsearch and Kibana
3. Create any alerting rule with a short interval (10s) and wait for it to run
4. Apply the following change to your code and restart Kibana
```diff --git a/x-pack/packages/kbn-alerting-state-types/src/task_state/v1/schema.ts b/x-pack/packages/kbn-alerting-state-types/src/task_state/v1/schema.ts
index 62e802483dc..3b4cd0025d1 100644
--- a/x-pack/packages/kbn-alerting-state-types/src/task_state/v1/schema.ts
+++ b/x-pack/packages/kbn-alerting-state-types/src/task_state/v1/schema.ts
@@ -42,6 +42,7 @@ export const rawAlertInstanceSchema = schema.object({
 });

 export const versionSchema = schema.object({
+  foo: schema.string(),
   alertTypeState: schema.maybe(ruleStateSchema),
   // tracks the active alerts
   alertInstances: schema.maybe(schema.recordOf(schema.string(), rawAlertInstanceSchema)),
```
5. Verify that you see the following logs and that task manager is still able to poll for tasks
```
[WARN ][plugins.taskManager] Task ... has a validation error: [foo]: expected value of type [string] but got [undefined]
[WARN ][plugins.taskManager] Task ... failed in attempt to run: [foo]: expected value of type [string] but got [undefined]
```
